src/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedView.Windows.cs

```diff
@@ -119,24 +118,25 @@ public void Flush(UIntPtr capacity)
                     // this strategy successfully flushed the view after no more than 3 retries.
 
                     int error = Marshal.GetLastWin32Error();
-                    bool canRetry = (!success && error == Interop.Errors.ERROR_LOCK_VIOLATION);
+                    if (error != Interop.Errors.ERROR_LOCK_VIOLATION)
+                        throw Win32Marshal.GetExceptionForWin32Error(error);
 
                     SpinWait spinWait = new SpinWait();
-                    for (int w = 0; canRetry && w < MaxFlushWaits; w++)
+                    for (int w = 0; w < MaxFlushWaits; w++)
                     {
                         int pause = (1 << w);  // MaxFlushRetries should never be over 30
                         MemoryMappedFile.ThreadSleep(pause);
 
-                        for (int r = 0; canRetry && r < MaxFlushRetriesPerWait; r++)
+                        for (int r = 0; r < MaxFlushRetriesPerWait; r++)
                         {
-                            success = Interop.Kernel32.FlushViewOfFile((IntPtr)firstPagePtr, capacity) != 0;
-                            if (success)
-                                return; // This will visit the finally block.
+                            if (Interop.Kernel32.FlushViewOfFile((IntPtr)firstPagePtr, capacity))
+                                return;
 
                             spinWait.SpinOnce();
 
                             error = Marshal.GetLastWin32Error();
-                            canRetry = (error == Interop.Errors.ERROR_LOCK_VIOLATION);
+                            if (error != Interop.Errors.ERROR_LOCK_VIOLATION)
+                                throw Win32Marshal.GetExceptionForWin32Error(error);
```

Comment:
As long as you're changing this, the SpinOnce should be moved down to after the error check.

Reply:
k